### PR TITLE
fix(trie): reveal extension child in sparse trie when updating a leaf

### PR DESCRIPTION
**Pull request by shekhirin**
Consider the following scenario:
1. We need to insert a leaf with the path `0x77853ecb3c80cc9dd0543d964574eefd6e89131a8e875c3606451dabd99ae6f4`
2. We query the proof, and reveal nodes down to `0x77853ec` which is an extension node `Extension(ExtensionNode { key: Nibbles("0d"), child: "a04d1721cf2c8ae2649fffbf4c5bc6e71ac1646e592ff4bfeb3054aa4d56b69456" })`
3. We get to inserting the leaf `0x77853ecb3c80cc9dd0543d964574eefd6e89131a8e875c3606451dabd99ae6f4` and turn the extension node into a branch node, setting the old nibble `d` from the extension node in the state mask of the new branch node.

The problem here is that the node under the nibble `d` may be a branch node that will need to appear in the hash mask of the newly created branch node, but we don't know if that's a branch node unless we reveal the extension node child.

---

This PR reveals the original extension node child when we turn an extension node into a branch node with two children.

### CODE REVIEW

File: crates/trie/sparse/src/trie.rs [hunk 1]
```diff
@@ -866,6 +774,110 @@ where
...
+                SparseNode::Extension { key, .. } => {
+                    current.extend_from_slice(key);
+
+                    if !path.starts_with(&current) {
+                        // find the common prefix
+                        let common = current.common_prefix_length(&path);
+                        *key = current.slice(current.len() - key.len()..common);
+
+                        // Check if the extension node child is a hash that needs to be revealed
+                        if self.nodes.get(&current).unwrap().is_hash() {
```
Comment (rkrasiuk) [inline crates/trie/sparse/src/trie.rs:836]: only if updates are enabled

Comment (shekhirin) [reply crates/trie/sparse/src/trie.rs]: done

File: crates/trie/sparse/src/trie.rs [hunk 2]
```diff
@@ -866,6 +774,116 @@ where
...
+                        // find the common prefix
+                        let common = current.common_prefix_length(&path);
+                        *key = current.slice(current.len() - key.len()..common);
+
+                        // If branch node updates retention is enabled, we need to query the
+                        // extension node child to later set the hash mask for a parent branch node
+                        // correctly.
+                        if self.updates.is_some() {
+                            // Check if the extension node child is a hash that needs to be revealed
+                            if self.nodes.get(&current).unwrap().is_hash() {
```
Comment (rkrasiuk) [inline crates/trie/sparse/src/trie.rs:840]: surprised clippy does not complain

**Review by rkrasiuk — APPROVED** [review]
lgtm

**Pull request merged**